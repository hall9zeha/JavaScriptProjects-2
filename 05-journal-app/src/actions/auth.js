// Las acciones no son más que simples funciones que representan nuestros casos de uso

import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/Types";
import { uiFinishLoading, uiStartLoading } from "./ui";



export const startLoginWithEmailPassword = (email, password) =>{
    return (dispatch)=>{
        dispatch(uiStartLoading())
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName));
                dispatch(uiFinishLoading())
            })
            .catch(e=>{
                dispatch(uiFinishLoading())
                console.log(e)
            })
        
    }
}

export const startLoginWithGoogle = ()=>{
    return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user})=>{
                dispatch(login(user.uid,user.displayName))
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password,name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then(async({user}) =>{
                await user.updateProfile({displayName:name});
                dispatch(login(user.uid,user.displayName));
            })
            .catch(e=>{
                console.log(e)
            })
    }
}


// La siguiente función retornará un objeto
export const login = (uid, displayName)=>({
    type: types.login,
    payload:{
        uid,
        displayName
    }
})

// Logout

export const startLogout = ()=>{
    return async (dispatch)=>{
        await firebase.auth().signOut();
        dispatch(logout())
    }
}

export const logout=()=>({
    type:types.logout
})