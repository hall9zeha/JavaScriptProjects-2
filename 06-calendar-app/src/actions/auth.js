import Swal from "sweetalert2";
import { fetchWithToken, fetchWithoutToken } from "../helpers/fetch"
import { types } from "../types/types";

export const startLogin = (email, password) =>{
    //Gracias a la librería thunk podemos disponer de un dispatch
    return async(dispatch) =>{
        const resp = await fetchWithoutToken('auth',{email,password},'POST');
        const body = await resp.json();
        
        //Si recibimos un arespuesta correcta de nuestra api 
        /*
            {
                ok:true
                ...
            }
        guardamos el token y la fecha de creación en local storage
        */
       
        if(body.ok){
            localStorage.setItem('token',body.token);
            localStorage.setItem('token-init-date',new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))
        }else{
            Swal.fire('Error',body.msg, 'error');
        }
    }
}

export const startRegister = (name,email,password) =>{
    return async(dispatch)=>{
        const resp = await fetchWithoutToken('auth/new',{name,email,password},'POST');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))

        }else{
            Swal.fire('Error',body.msg, 'error');
        }

    }
}
export const startChecking =()=> {
    return async(dispatch)=>{

        //No requiere ningún argumento porque por defecto será una petición GET
        const resp = await fetchWithToken('auth/renew');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid:body.uid,
                name:body.name
            }))

        }else{
            Swal.fire('Error',body.msg, 'error');
            dispatch(checkingFinish());
        }
    }
}

const checkingFinish = () =>({type:types.authCheckingFinish});
//Ya que usaremos esta acción solo en este lugar no la exportaremos
const login = (user) =>({
    type:types.authLogin,
    payload:user
})

