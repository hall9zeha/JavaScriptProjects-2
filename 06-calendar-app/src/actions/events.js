import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch"
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types"

//Si se guardó nuestro evento en la base de datos mongo en la nube entonces disparar eventAddNew()
export const eventStartAddNew = (event)=>{
    return async(dispatch, getState)=>{
        //Obtenemos el nombre, uid del usuario que ya está en nuestro store de redux
        const {uid, name} = getState().auth;
        try {
            
            const resp = await fetchWithToken('events',event,'POST');
            const body = await resp.json();
            console.log(body)
            if(body.ok){
                event.id=body.savedEvent.id;
                event.user = {
                    _id:uid,
                    name:name
                }
                
                dispatch(eventAddNew(event))
            }
        
        } catch (error) {
            console.log(error)
        }
    }
}

const eventAddNew =(event)=>({
    type:types.eventAddNew,
    payload:event
})

export const eventSetActive =(event)=>({
    type:types.eventSetActive,
    payload:event
})
export const eventClearActiveEvent = () =>({type:types.eventClearActiveEvent})

export const eventStartUpdate =(event) =>{
    return async(dispatch)=>{
        try {
            const resp = await fetchWithToken(`events/${event.id}`,event,'PUT');
            const body = await resp.json();

            if(body.ok){
                dispatch(eventUpdated(event));
            }else{
                Swal.fire('Error',body.msg,'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
}

const eventUpdated =(event)=>({
    type:types.eventUpdated,
    payload:event
})

export const eventStartDelete =()=>{
    return async(dispatch,getState)=>{
        const {id} = getState().calendar.activeEvent;
        
        try {
            const resp = await fetchWithToken(`events/${id}`,{},'DELETE');
            const body = await resp.json();

            if(body.ok){
                dispatch(eventDeleted());
              
            }else{
                Swal.fire('Error',body.msg,'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
}

const eventDeleted = ()=>({
    type:types.eventDeleted
})

export const eventStartLoading = ()=>{

    return async(dispatch,getState)=>{
        const {uid} = getState().auth;
        try {
            const resp = await fetchWithToken('events');
            const body = await resp.json();
            const events = prepareEvents(uid,body.events);
            dispatch(eventLoaded(events));

        } catch (error) {
            
        }
    }
}

const eventLoaded=(events)=>({
    type:types.eventLoaded,
    payload:events
})

export const eventLogout=()=>({
    type:types.eventLogout
})