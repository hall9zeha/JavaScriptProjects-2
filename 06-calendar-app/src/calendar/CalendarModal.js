import moment from 'moment';
import React, { useState } from 'react'

import 'react-datetime-picker/dist/DateTimePicker.css';
import 'react-calendar/dist/Calendar.css';
import 'react-clock/dist/Clock.css';
import Modal from 'react-modal'
import DateTimePicker from 'react-datetime-picker';
import Swal from 'sweetalert2';


const customStyles = {
    content: {
      top: '50%',
      left: '50%',
      right: 'auto',
      bottom: 'auto',
      marginRight: '-50%',
      transform: 'translate(-50%, -50%)',
    },
  };

Modal.setAppElement('#root');
const nowDate =  moment().minutes(0).seconds(0).add(1,'hours')
const nowDatePlus = nowDate.clone().add(1,'hours')


export const CalendarModal = () => {
  const [dateStart, setDateStart] = useState(nowDate.toDate())
  const [dateEnd, setDateEnd] = useState(nowDatePlus.toDate())
  const [titleValid, setTitleValid] = useState(true)
  
  const [formValues, setFormValues] = useState({
    title:'Event',
    notes:'',
    start:nowDate.toDate(),
    end:nowDatePlus.toDate()
  })
  const {notes,title, start,end} = formValues;

  const handleInputchange = ({target})=>{
    setFormValues({
      ...formValues,
      [target.name]:target.value
    })
  }

  const closeModal = ()=>{
       
    }
    //Al escoger una fecha en el datepicker la capturamos con esta función
    const handleStartDateChange = (e)=>{
      setDateStart(e)
      setFormValues({
        ...formValues,
        start:e
      })
    }
    const handleEndDateChange = (e)=>{
      setDateEnd(e)
      setFormValues({
        ...formValues,
        end:e
      })
    }
    const handleSubmitForm = (e)=>{
      e.preventDefault()
      const momentStart = moment(start)
      const momentEnd = moment(end)
      //Validamos que la fecha inicial no sea igual o mayor que la fecha final
      if(momentStart.isSameOrAfter(momentEnd)){
        Swal.fire('Error','La fecha fin debe ser mayor que la fecha de inicio')  
        return;
      }
      if(title.trim().length <2){
        return setTitleValid(false)
      }
      setTitleValid(true)
      closeModal()
    }

  return (
    <Modal
        isOpen={true}
        // onAfterOpen={afterOpenModal}
        onRequestClose={closeModal}
        style={customStyles}
        className='modal'
        overlayClassName='modal-fondo'
        closeTimeoutMS={200}
        >
        <h1> Nuevo evento </h1>
        <hr />
        <form className="container"
              onSubmit={handleSubmitForm}
        >

            <div className="form-group mb-2">
                <label>Fecha y hora inicio</label>
                <DateTimePicker
                    onChange={handleStartDateChange}
                    value={dateStart}
                    className='form-control'/>
            </div>

            <div className="form-group mb-2">
                <label>Fecha y hora fin</label>
                <DateTimePicker
                    onChange={handleEndDateChange}
                    value={dateEnd}
                    //La fecha inicial no debe ser mayor que dateEnd, lo validamos con minDate 
                    minDate={dateStart}
                    className='form-control'/>
            </div>

            <hr />
            <div className="form-group mb-2">
                <label>Titulo y notas</label>
                <input 
                    type="text" 
                    // Si nuestro campo de texto está vacío nos mostrará una alerta con la clase is-invalid
                    className={`form-control ${!titleValid && 'is-invalid'}`}
                    placeholder="Título del evento"
                    name="title"
                    value={title}
                    onChange={handleInputchange}
                    autoComplete="off"
                />
                <small id="emailHelp" className="form-text text-muted">Una descripción corta</small>
            </div>

            <div className="form-group mb-2">
                <textarea 
                    type="text" 
                    className="form-control"
                    placeholder="Notas"
                    rows="5"
                    name="notes"
                    value={notes}
                    onChange={handleInputchange}
                ></textarea>
                <small id="emailHelp" className="form-text text-muted">Información adicional</small>
            </div>

            <button
                type="submit"
                className="btn btn-outline-primary btn-block"
            >
                <i className="far fa-save"></i>
                <span> Guardar</span>
            </button>

        </form>
    </Modal>
  )
}
