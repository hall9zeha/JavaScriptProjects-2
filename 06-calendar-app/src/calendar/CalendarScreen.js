import React from 'react'
import {Calendar,momentLocalizer} from 'react-big-calendar'
import { Navbar } from '../ui/Navbar'
import moment from 'moment'
import 'react-big-calendar/lib/css/react-big-calendar.css' //Estilos para big calendar

const localizer = momentLocalizer(moment)
const events =[{
    title:'CumpleaÃ±os de Martha',
    start:moment().toDate(),//como hacer new Date()
    end:moment().add(2,'hours').toDate()//Agregamos dos horas despues de la fecha inicial
}]
export const CalendarScreen = () => {
  return (
    <div className='calendar-screen'>
        <Navbar/>
        <Calendar
            localizer={localizer}
            events={events}
            starAccessor="start"
            endAccessor="end"
        />
    </div>
  )
}
