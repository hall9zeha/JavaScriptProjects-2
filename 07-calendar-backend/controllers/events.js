const Event = require("../models/Event")

//Get events
const getEvents = async (req, res)=>{

    const events = await Event.find()
    //Rellenamos los datos de usuario correspondiente a la referencia por id que está en nuestra propiedad 'user'
    //esto lo hará mongoose por nosotros, podemos especificar los campos que queremos, en este caso ('name') ex: 'name password', sin comas y dentro de la misma cadena
    //si solo ponemos el nombre de la referencia (user) traerá todos los campos
                            .populate('user','name');
    res.status(200).json({
        ok:true,
        msg:"Eventos cargados",
        events
    })
}

//Create event
const createEvent = async(req,res)=>{
    
    const event = new Event(req.body);
    event.user = req.uid;
    try {
        const savedEvent = await event.save();
        res.json({
            ok:true,
            savedEvent
        })
        
    } catch (error) {
        console.log(error);
        res.status(500).json({
            of:false,
            msg:'Error creating event'
        })
    }
   
}

//Update event
const updateEvent =async (req,res)=>{
    //Obtenemos el id del registro enviado a través de la url
    const eventId = req.params.id;

    try {
        //Si el evento existe
        const event = await Event.findById(eventId)
        if(!event){
            return res.status(404).json({
                ok:false,
                msg:'Event not exist'
            })
        }
        //Si el usuario es el mismo que creó el evento permitir caso contrario no
        //Donde event.user ('user') es la referencia que es un 'id'
        if(event.user.toString() !== req.uid){
            return res.status(401).json({
                ok:false,
                msg:"You don't have privileges to modify this event"
            })
        }
        const eventForUpdate = {
            ...req.body,
            user:req.uid
        };
        
        //Por defecto findByIdAndUpdate retorna el último registro sin actualizar por si se requiere
        //Entonces debemos poner {new:true} como tercer argumento para que nos devuelva el registro actualizado
        const eventUpdated = await Event.findByIdAndUpdate(eventId,eventForUpdate,{new:true});

        res.json({
            ok:true,
            eventUpdated
        })

    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok:false,
            msg:'Error updating event'
        })
    }
   
}

//Delete event
const deleteEvent = async(req,res)=>{
     //Obtenemos el id del registro enviado a través de la url
     const eventId = req.params.id;

     try {
         //Si el evento existe
         const event = await Event.findById(eventId)
         if(!event){
             return res.status(404).json({
                 ok:false,
                 msg:'Event not exist'
             })
         }
         //Si el usuario es el mismo que creó el evento permitir caso contrario no
         //Donde event.user ('user') es la referencia que es un 'id'
         if(event.user.toString() !== req.uid){
             return res.status(401).json({
                 ok:false,
                 msg:"You don't have privileges to delete this event"
             })
         }
        
         //Por defecto findByIdAndUpdate retorna el último registro sin actualizar por si se requiere
         //Entonces debemos poner {new:true} como tercer argumento para que nos devuelva el registro actualizado
         const eventDeleted= await Event.findByIdAndDelete(eventId);
 
         res.json({
             ok:true,
             eventDeleted
         })
 
     } catch (error) {
         console.log(error);
         res.status(500).json({
             ok:false,
             msg:'Error delete event'
         })
     }
    
}

module.exports = {
    getEvents, createEvent, updateEvent, deleteEvent
}