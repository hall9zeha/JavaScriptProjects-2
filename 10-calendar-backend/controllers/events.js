const Event = require("../models/Event")

//Get events
const getEvents = (req, res)=>{
    res.status(200).json({
        ok:true,
        msg:"Eventos cargados"
    })
}

//Create event
const createEvent = async(req,res)=>{
    
    const event = new Event(req.body);
    event.user = req.uid;
    try {
        const savedEvent = await event.save();
        res.json({
            ok:true,
            savedEvent
        })
        
    } catch (error) {
        console.log(error);
        res.status(500).json({
            of:false,
            msg:'Error creating event'
        })
    }
   
}

//Update event
const updateEvent = (req,res)=>{
    res.status(200).json({
        ok:true,
        msg:"Evento actualizado"
    })
}

//Delete event
const deleteEvent = (req,res)=>{
    res.status(200).json({
        ok:true,
        msg:"Evento eliminado"
    })
}

module.exports = {
    getEvents, createEvent, updateEvent, deleteEvent
}