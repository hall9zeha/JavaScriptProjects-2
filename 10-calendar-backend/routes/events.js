
const {Router} = require('express');
const { getEvents, createEvent, updateEvent, deleteEvent } = require('../controllers/events');
const { validateJWT } = require('../middlewares/jws-validator');
const { check } = require('express-validator');
const { fieldValidate } = require('../middlewares/field-validators');
const { isDate } = require('../helpers/isDate');
const router = Router();


/*
Al utilizar validateJWT en cada ruta de la siguiente manera:
ex: router.get('/',validate,getEvents), haríamos esto en cuatro lugares.
Tenemos la opción de declarar esta acción de validar token una sola vez para que se ejecute antes de cada ruta llamada
*/
router.use(validateJWT);


//Get events

router.get('/', getEvents);

//Create new event
router.post('/',
    [
        check('title','The title is required').not().isEmpty(),
        check('start','Initial date is required').custom(isDate),
        check('end','End date is required').custom(isDate),
        fieldValidate
    ],
    createEvent);

//Update event
router.put('/:id',updateEvent);

//Delete event
router.delete('/:id',deleteEvent);


module.exports = router;
