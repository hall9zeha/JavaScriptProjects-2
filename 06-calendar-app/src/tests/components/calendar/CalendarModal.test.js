import React from 'react'
import {Provider} from 'react-redux';
import renderer ,{act}from 'react-test-renderer';
import ReactDOM from 'react-dom';

import configureStore from 'redux-mock-store'
import { CalendarModal } from '../../../components/calendar/CalendarModal';
import moment from 'moment';
import { eventStartUpdate,eventClearActiveEvent } from '../../../actions/events';



jest.mock('../../../actions/events',()=>({
  eventStartUpdate:jest.fn(),
  eventClearActiveEvent:jest.fn()
}))

//Si usamos react-renderer y queremos hacer snapshots en componentes modales
//debemos burlar(hacer un mock) a react doom
jest.mock('react-dom', () => {
    const original = jest.requireActual('react-dom');
      return {
      ...original,
      createPortal: (node) => node,
    };
  });

const thunk = require('redux-thunk').thunk;


//***************************/
const middlewares = [thunk]
const mockStore = configureStore(middlewares)



const nowDate =  moment().minutes(0).seconds(0).add(1,'hours')
const nowDatePlus = nowDate.clone().add(1,'hours')

const initState={
    ui:{
        modalOpen:true,
    },
    calendar:{
        events:[],
        activeEvent:{
            title:"test title",
            notes:"test note",
            start:nowDate.toDate(),
            end:nowDatePlus.toDate(),
            id:'123'
        }
    },
    auth:{ uid:'123456789', name:'Barry'}
}

const store =   mockStore(initState)
store.dispatch = jest.fn()

const wrapper =  renderer.create(
    <Provider store={store}>
           <CalendarModal/>
    </Provider>
)


describe('Tests in <CalendarModal/> component', () => { 
    //El modal no se puede renderizar sin simular lo siguiente
    // beforeAll(() => {
    //     ReactDOM.createPortal = jest.fn((element, node) => {
    //       return element
    //     })
    //   })
  
    afterEach(() => {
      //ReactDOM.createPortal.mockClear()
      jest.clearAllMocks()
    })

    test('should show successfully', () => { 
        expect(wrapper).toMatchSnapshot()
      
     })
     test('should call update action and close modal', () => { 
      act(()=>{
        wrapper.root.findByType('form').props.onSubmit({preventDefault:jest.fn()})
      })
        
        expect(eventStartUpdate).toHaveBeenCalledWith({
          title:"test title",
          notes:"test note",
          start:nowDate.toDate(),
          end:nowDatePlus.toDate(),
          id:'123'
      })
        expect(eventClearActiveEvent).toHaveBeenCalled()
      })
      test('should show error if title is empty', () => { 
        act(()=>{
          wrapper.root.findByType('form').props.onSubmit({preventDefault:jest.fn()})
        })
        
        expect(wrapper.root.findByProps({className:'form-control is-invalid'})).toBeTruthy()
       })
 })