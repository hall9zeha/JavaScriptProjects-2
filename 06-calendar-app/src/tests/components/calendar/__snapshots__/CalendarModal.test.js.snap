// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <CalendarModal/> component should show successfully 1`] = `
<div
  className="ReactModal__Overlay modal-fondo"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={Object {}}
>
  <div
    aria-modal={true}
    className="ReactModal__Content modal"
    data-testid="modal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={
      Object {
        "bottom": "auto",
        "left": "50%",
        "marginRight": "-50%",
        "right": "auto",
        "top": "50%",
        "transform": "translate(-50%, -50%)",
      }
    }
    tabIndex="-1"
  >
    <h1>
       
      Editar evento
       
    </h1>
    <hr />
    <form
      className="container"
      onSubmit={[Function]}
    >
      <div
        className="form-group mb-2"
      >
        <label>
          Fecha y hora inicio
        </label>
        <div
          className="react-datetime-picker react-datetime-picker--closed react-datetime-picker--enabled form-control"
          onFocus={[Function]}
        >
          <div
            className="react-datetime-picker__wrapper"
          >
            <div
              className="react-datetime-picker__inputGroup"
              onClick={[Function]}
            >
              <input
                hidden={true}
                max="275760-09-12T19:00"
                min="0001-01-01T00:00"
                name="datetime"
                onChange={[Function]}
                onFocus={[Function]}
                step={60}
                style={
                  Object {
                    "position": "absolute",
                    "visibility": "hidden",
                    "zIndex": "-999",
                  }
                }
                type="datetime-local"
                value=""
              />
              <input
                autoComplete="off"
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__month"
                data-input="true"
                inputMode="numeric"
                max={12}
                min={1}
                name="month"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={false}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                /
              </span>
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__day"
                data-input="true"
                inputMode="numeric"
                max={31}
                min={1}
                name="day"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={false}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                /
              </span>
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__year"
                data-input="true"
                inputMode="numeric"
                max={275760}
                min={1}
                name="year"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="----"
                required={false}
                step={1}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                Â 
              </span>
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__hour"
                data-input="true"
                inputMode="numeric"
                max={12}
                min={1}
                name="hour12"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={false}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                :
              </span>
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__minute"
                data-input="true"
                inputMode="numeric"
                max={59}
                min={0}
                name="minute"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={false}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                 
              </span>
              <select
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__amPm"
                data-input="true"
                data-select="true"
                name="amPm"
                onChange={[Function]}
                onKeyDown={[Function]}
                required={false}
                value=""
              >
                <option
                  value=""
                >
                  --
                </option>
                <option
                  disabled={false}
                  value="am"
                >
                  AM
                </option>
                <option
                  disabled={false}
                  value="pm"
                >
                  PM
                </option>
              </select>
            </div>
            <button
              className="react-datetime-picker__clear-button react-datetime-picker__button"
              onClick={[Function]}
              onFocus={[Function]}
              type="button"
            >
              <svg
                className="react-datetime-picker__clear-button__icon react-datetime-picker__button__icon"
                height={19}
                stroke="black"
                strokeWidth={2}
                viewBox="0 0 19 19"
                width={19}
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="4"
                  x2="15"
                  y1="4"
                  y2="15"
                />
                <line
                  x1="15"
                  x2="4"
                  y1="4"
                  y2="15"
                />
              </svg>
            </button>
            <button
              className="react-datetime-picker__calendar-button react-datetime-picker__button"
              onClick={[Function]}
              onFocus={[Function]}
              type="button"
            >
              <svg
                className="react-datetime-picker__calendar-button__icon react-datetime-picker__button__icon"
                height={19}
                stroke="black"
                strokeWidth={2}
                viewBox="0 0 19 19"
                width={19}
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="none"
                  height="15"
                  width="15"
                  x="2"
                  y="2"
                />
                <line
                  x1="6"
                  x2="6"
                  y1="0"
                  y2="4"
                />
                <line
                  x1="13"
                  x2="13"
                  y1="0"
                  y2="4"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="form-group mb-2"
      >
        <label>
          Fecha y hora fin
        </label>
        <div
          className="react-datetime-picker react-datetime-picker--closed react-datetime-picker--enabled form-control"
          onFocus={[Function]}
        >
          <div
            className="react-datetime-picker__wrapper"
          >
            <div
              className="react-datetime-picker__inputGroup"
              onClick={[Function]}
            >
              <input
                hidden={true}
                max="275760-09-12T19:00"
                min="2024-03-28T18:00"
                name="datetime"
                onChange={[Function]}
                onFocus={[Function]}
                step={60}
                style={
                  Object {
                    "position": "absolute",
                    "visibility": "hidden",
                    "zIndex": "-999",
                  }
                }
                type="datetime-local"
                value=""
              />
              <input
                autoComplete="off"
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__month"
                data-input="true"
                inputMode="numeric"
                max={12}
                min={1}
                name="month"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={false}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                /
              </span>
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__day"
                data-input="true"
                inputMode="numeric"
                max={31}
                min={1}
                name="day"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={false}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                /
              </span>
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__year"
                data-input="true"
                inputMode="numeric"
                max={275760}
                min={2024}
                name="year"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="----"
                required={false}
                step={1}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                Â 
              </span>
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__hour"
                data-input="true"
                inputMode="numeric"
                max={12}
                min={1}
                name="hour12"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={false}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                :
              </span>
              <input
                autoComplete="off"
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__minute"
                data-input="true"
                inputMode="numeric"
                max={59}
                min={0}
                name="minute"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder="--"
                required={false}
                type="number"
                value=""
              />
              <span
                className="react-datetime-picker__inputGroup__divider"
              >
                 
              </span>
              <select
                autoFocus={false}
                className="react-datetime-picker__inputGroup__input react-datetime-picker__inputGroup__amPm"
                data-input="true"
                data-select="true"
                name="amPm"
                onChange={[Function]}
                onKeyDown={[Function]}
                required={false}
                value=""
              >
                <option
                  value=""
                >
                  --
                </option>
                <option
                  disabled={false}
                  value="am"
                >
                  AM
                </option>
                <option
                  disabled={false}
                  value="pm"
                >
                  PM
                </option>
              </select>
            </div>
            <button
              className="react-datetime-picker__clear-button react-datetime-picker__button"
              onClick={[Function]}
              onFocus={[Function]}
              type="button"
            >
              <svg
                className="react-datetime-picker__clear-button__icon react-datetime-picker__button__icon"
                height={19}
                stroke="black"
                strokeWidth={2}
                viewBox="0 0 19 19"
                width={19}
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="4"
                  x2="15"
                  y1="4"
                  y2="15"
                />
                <line
                  x1="15"
                  x2="4"
                  y1="4"
                  y2="15"
                />
              </svg>
            </button>
            <button
              className="react-datetime-picker__calendar-button react-datetime-picker__button"
              onClick={[Function]}
              onFocus={[Function]}
              type="button"
            >
              <svg
                className="react-datetime-picker__calendar-button__icon react-datetime-picker__button__icon"
                height={19}
                stroke="black"
                strokeWidth={2}
                viewBox="0 0 19 19"
                width={19}
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="none"
                  height="15"
                  width="15"
                  x="2"
                  y="2"
                />
                <line
                  x1="6"
                  x2="6"
                  y1="0"
                  y2="4"
                />
                <line
                  x1="13"
                  x2="13"
                  y1="0"
                  y2="4"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <hr />
      <div
        className="form-group mb-2"
      >
        <label>
          Titulo y notas
        </label>
        <input
          autoComplete="off"
          className="form-control false"
          name="title"
          onChange={[Function]}
          placeholder="TÃ­tulo del evento"
          type="text"
          value="test title"
        />
        <small
          className="form-text text-muted"
          id="emailHelp"
        >
          Una descripciÃ³n corta
        </small>
      </div>
      <div
        className="form-group mb-2"
      >
        <textarea
          className="form-control"
          name="notes"
          onChange={[Function]}
          placeholder="Notas"
          rows="5"
          type="text"
          value="test note"
        />
        <small
          className="form-text text-muted"
          id="emailHelp"
        >
          InformaciÃ³n adicional
        </small>
      </div>
      <button
        className="btn btn-outline-primary btn-block"
        type="submit"
      >
        <i
          className="far fa-save"
        />
        <span>
           Guardar
        </span>
      </button>
    </form>
  </div>
</div>
`;
